File: puppy_camera/scripts/camera_node.py
#!/usr/bin/env python3

import rospy
import cv2
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
import numpy as np

class CameraNode:
    def __init__(self):
        # Initialize the ROS node
        rospy.init_node('camera_node', anonymous=True)
        
        # Get parameters
        self.camera_device_id = rospy.get_param('~camera_device_id', 0)
        self.frame_rate = rospy.get_param('~frame_rate', 30)
        self.frame_id = rospy.get_param('~frame_id', 'camera_link')
        
        # Initialize OpenCV video capture
        self.cap = cv2.VideoCapture(self.camera_device_id)
        if not self.cap.isOpened():
            rospy.logerr(f"Could not open camera with device ID {self.camera_device_id}")
            return
        
        # Set camera properties
        self.cap.set(cv2.CAP_PROP_FPS, self.frame_rate)
        
        # Initialize the CvBridge
        self.bridge = CvBridge()
        
        # Create an image publisher
        self.image_pub = rospy.Publisher('camera/image_raw', Image, queue_size=1)
        
        rospy.loginfo("Camera node started. Publishing to camera/image_raw topic.")
        
        # Create a timer to capture and publish frames
        self.timer = rospy.Timer(rospy.Duration(1.0/self.frame_rate), self.capture_and_publish)
        
    def capture_and_publish(self, event):
        # Capture a frame
        ret, frame = self.cap.read()
        
        if ret:
            try:
                # Convert OpenCV image to ROS message
                img_msg = self.bridge.cv2_to_imgmsg(frame, "bgr8")
                img_msg.header.stamp = rospy.Time.now()
                img_msg.header.frame_id = self.frame_id
                
                # Publish the image
                self.image_pub.publish(img_msg)
            except CvBridgeError as e:
                rospy.logerr(f"CvBridge Error: {e}")
    
    def __del__(self):
        # Clean up
        if hasattr(self, 'cap') and self.cap.isOpened():
            self.cap.release()

if __name__ == '__main__':
    try:
        camera_node = CameraNode()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: puppy_camera/src/camera_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <image_transport/image_transport.h>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>

int main(int argc, char** argv)
{
    // Initialize the ROS node
    ros::init(argc, argv, "camera_node");
    ros::NodeHandle nh;
    ros::NodeHandle pnh("~");
    
    // Get parameters
    int camera_device_id;
    int frame_rate;
    std::string frame_id;
    
    pnh.param("camera_device_id", camera_device_id, 0);
    pnh.param("frame_rate", frame_rate, 30);
    pnh.param("frame_id", frame_id, std::string("camera_link"));
    
    // Create image transport publisher
    image_transport::ImageTransport it(nh);
    image_transport::Publisher image_pub = it.advertise("camera/image_raw", 1);
    
    // Open the camera
    cv::VideoCapture cap(camera_device_id);
    if (!cap.isOpened()) {
        ROS_ERROR("Could not open camera with device ID %d", camera_device_id);
        return -1;
    }
    
    // Set camera properties if needed
    cap.set(cv::CAP_PROP_FPS, frame_rate);
    
    // Create rate limiter
    ros::Rate loop_rate(frame_rate);
    
    // Main loop
    cv::Mat frame;
    sensor_msgs::ImagePtr msg;
    
    ROS_INFO("Camera node started. Publishing to camera/image_raw topic.");
    
    while (ros::ok()) {
        // Capture frame
        cap >> frame;
        
        if (!frame.empty()) {
            // Convert OpenCV image to ROS message
            msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", frame).toImageMsg();
            msg->header.stamp = ros::Time::now();
            msg->header.frame_id = frame_id;
            
            // Publish the image
            image_pub.publish(msg);
        }
        
        ros::spinOnce();
        loop_rate.sleep();
    }
    
    return 0;
}
